ocaml/typing/includecore.ml

```diff
@@ -543,11 +543,20 @@ module Record_diffing = struct
   let compare_labels env params1 params2
         (ld1 : Types.label_declaration)
         (ld2 : Types.label_declaration) =
-        if ld1.ld_mutable <> ld2.ld_mutable
-        then
-          let ord = if ld1.ld_mutable = Asttypes.Mutable then First else Second in
-          Some (Mutability ord)
-        else begin
+        let mut =
+          match ld1.ld_mutable, ld2.ld_mutable with
+          | Immutable, Immutable -> None
+          | Mutable _, Immutable -> Some First
+          | Immutable, Mutable _ -> Some Second
+          | Mutable m1, Mutable m2 ->
+              if Misc.eq_from_le Mode.Alloc.Const.le m1 m2 then None
+              else
+                Misc.fatal_errorf "Unexpected mutable(%a) = mutable(%a)"
+                  Mode.Alloc.Const.print m1 Mode.Alloc.Const.print m2
```

Comment:
If you're going to error anyway, it's probably better to just error if either of the modes isn't legacy, rather than letting it succeed in some invalid cases but not others.